llvm/lib/LTO/LTO.cpp

```diff
@@ -626,7 +636,12 @@ void LTO::addModuleToGlobalRes(ArrayRef<InputFile::Symbol> Syms,
     assert(ResI != ResE);
     SymbolResolution Res = *ResI++;
 
-    auto &GlobalRes = (*GlobalResolutions)[Sym.getName()];
+    StringRef SymbolName = Sym.getName();
+    // Keep copies of symbols if the client of LTO says so.
+    if (GlobalResolutionSymbolSaver && !GlobalResolutions->contains(SymbolName))
+      SymbolName = GlobalResolutionSymbolSaver->save(SymbolName);
+
+    auto &GlobalRes = (*GlobalResolutions)[SymbolName];
```

Comment:
A somewhat nit picky suggestion, but should we create a wrapper class for `llvm::DenseMap<StringRef, GlobalResolution>`?  This way, we can intercept its `operator()[]` and put all the deduplication logic there.

This way, the application logic should stay clear.  That is:

```
auto &GlobalRes = (*GlobalResolutions)[Sym.getName()];
```

staying as is.

We should be able to put members like:

```
  std::unique_ptr<llvm::BumpPtrAllocator> Alloc;

  // Symbol saver for global resolution map.
  std::unique_ptr<llvm::StringSaver> GlobalResolutionSymbolSaver;
```

into the wrapper class and probably strip away `std::unique_ptr` from them, which in turn should let you remove:

```
  GlobalResolutionSymbolSaver.reset();
  Alloc.reset();
```

below.

Reply:
I added a helper method `LTO::releaseGlobalResolutionsMemory` which releases the memory used by GlobalResolutions together. I'm not sure if having a wrapper class has the correct ROI for code simplicity here as reader also need to parse the added logic of wrapper class.